### PR TITLE
Unify property extractor function names

### LINKED ISSUE #521: Consider renaming/unifying `prop`/`extracting`/`suspendCall`
`prop` isn't short for `property`–at least not in the Kotlin-sense. Its `KFunction` and lambda-accepting overloads let it operate on more than just Kotlin properties. It may be short for "property" in the general sense, but that's then confusing against the Kotlin concept and abbreviations are generally frowned upon in API signatures.

`extracting` is `prop` projected over a collection.

`suspendCall` is the lambda-accepting `prop` but `suspend`able.

These are all the same thing, and it would be cool if they had a unified naming convention.

I don't have any perfect naming proposals yet but I've been sitting on filing this issue for months now so I'm firing it off in the hopes others can help.

I came up with two ideas:

Center around the verb "have":
- Assert that some user has (a) name (that) is equal to "Alice":

  ```kotlin
  assertThat(someUser).has(User::name).isEqualTo("Alice")
  
  assertThat(someUser).has("name", { it.name() }).isEqualTo("Alice")
  ```

- Assert that all users each has (a) name (that) is equal to "Alice", "Bob", "Eve":

  ```kotlin
  assertThat(allUsers)
      .eachHas(User::name)
      .isEqualTo(listOf("Alice", "Bob", "Eve"))
  ```

  Is this `eachHave`? Tempted to bend grammar even if it is for consistency.

- Assert that some user has (a) DB name (that) is equal to "Alice":

  ```kotlin
  assertThat(someUser).has("name", { it.dbName() }).isEqualTo("Alice")
  ```

  Can we pull this off, where the signature matches? Can the main one be `inline` with some API changes maybe?

Alternatively, lean into `map` whose semantics are hopefully known from its available on `Iterable`s, `Flow`s, Rx, other languages, etc.

- Assert that some user map(ped to a) name (that) is equal to "Alice":

  ```kotlin
  assertThat(someUser).map(User::name).isEqualTo("Alice")
  
  assertThat(someUser).map("name", { it.name() }).isEqualTo("Alice")
  ```

- Assert that all users map(ped to their) name (that) is equal to "Alice", "Bob", "Eve":

  ```kotlin
  assertThat(allUsers)
      .map(User::name)
      .isEqualTo(listOf("Alice", "Bob", "Eve"))
  ```

  Is this `eachHave`? Tempted to bend grammar even if it is for consistency.

- Assert that some user map(ped to a) DB name (that) is equal to "Alice":

  ```kotlin
  assertThat(someUser).map("name", { it.dbName() }).isEqualTo("Alice")
  ```

Thoughts?

(Also whoops cmd+enter too soon)

### PR DESCRIPTION
**Pull request by grodin**
Closes #521

I'm opening this as a draft as the docs aren't updated yet and this is just supposed to be to get the bikeshedding going.

Known things to fix before merge:

- [ ] ~update docs~
- [x] update README
- [x] update CHANGELOG
- [x] update names of tests
- [x] deprecate old functions

### CODE REVIEW

**Comment by evant** [comment]
While we are still bike-shedding the naming I want to call out we'll want to keep but deprecate the old names, not get rid of them entirely.

**Comment by grodin** [comment]
Yeah, I realised that after I pushed those commits. I'll add that to the to-do list above.

**Comment by grodin** [comment]
Let me know if you want this squashed

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Changed
+- renamed `prop` to `having` as part of effort to unify API naming, old name is deprecated.
+- renamed `suspendCall` to `having` as part of effort to unify API naming, old name is deprecated.
+- renamed `extracting` to `eachHaving` as part of effort to unify API naming, old name is deprecated.
```
[the remark this answers is not in the document]
Comment (grodin) [reply CHANGELOG.md:12]: Just to clarify, you mean revert renaming `extracting` to `eachHaving`?

File: docs/assertk-coroutines/assertk.coroutines.assertions/contains-all.html [hunk 2]
```diff
@@ -41,7 +41,7 @@
 <a class="library-name--link" href="../../index.html">
                             assertk-project
                     </a>            </div>
-            <div class="library-version">0.28.0
```
[the remark this answers is not in the document]
Comment (grodin) [reply docs/assertk-coroutines/assertk.coroutines.assertions/contains-all.html]: https://github.com/willowtreeapps/assertk/pull/524/commits/a003cd240e4b113e7505aa68065d60c963647a90

File: CHANGELOG.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (grodin) [reply CHANGELOG.md]: If so, reverted in https://github.com/willowtreeapps/assertk/pull/524/commits/f2f2e21c27717ef13b5d12cc1bb29eb497234242

[the remark this answers is not in the document]
Comment (evant) [reply CHANGELOG.md]: In this pr yeah

File: docs/assertk-coroutines/assertk.coroutines.assertions/contains-all.html (hunk 2 above)
[the remark this answers is not in the document]
Comment (evant) [reply docs/assertk-coroutines/assertk.coroutines.assertions/contains-all.html]: and nvm I'm going to go ahead and update the docs

File: README.md [hunk 3]
```diff
@@ -148,12 +148,12 @@ semantics as follows:
 There's a few ways you extract the data you want to assert on. While you can do this yourself before calling the
 assertion, these methods will add the extra context to the failure message which can be helpful.
 
-The simplest way is with `prop()`. It will take a property (or function, or a name and a lambda) and return an
+The simplest way is with `having()`. It will take a property (or function, or a name and a lambda) and return an
```
Comment (evant) [inline README.md:151]: so the downside of updating the README now is that it'll show up on the main github page even though this change hasn't landed in a release yet, so I think we'll have to wait to change it unfortunately.

**Review by evant — COMMENTED** [review]
other than that looks good! While this should be squashed up to you if you want to do that, I can also do it while merging.

File: README.md (hunk 3 above)
Comment (grodin) [reply README.md]: Oops, oh yeah.

Reverted in https://github.com/willowtreeapps/assertk/pull/524/commits/3af4e67fd98dd40013653663aca389c41c7b7a8a

**Comment by grodin** [comment]
I'll let you squash it

File: assertk-coroutines/src/commonMain/kotlin/assertk/coroutines/assertions/any.kt [hunk 4]
```diff
@@ -10,8 +10,16 @@ import assertk.assertions.support.appendName
  * @param extract The suspend function to extract the property value out of the value of the current assert.
  *
  * ```
- * assertThat(person).suspendCall("name()", { it.name() }).isEqualTo("Sue")
+ * assertThat(person).having("name()", { it.name() }).isEqualTo("Sue")
  * ```
  */
+suspend fun <T, P> Assert<T>.having(name: String, extract: suspend (T) -> P): Assert<P> =
```
Comment (JakeWharton) [inline assertk-coroutines/src/commonMain/kotlin/assertk/coroutines/assertions/any.kt:16]: If we make the other one `inline`, we can eliminate this overload. Do we want that? Both APIs are public.

Comment (evant) [reply assertk-coroutines/src/commonMain/kotlin/assertk/coroutines/assertions/any.kt:16]: Yeah doing that would be fine

**Pull request merged**